### PR TITLE
authorisation - revoke privileges from group + delete member from group

### PR DESCRIPTION
**Pull request by iaincrawf**

### CODE REVIEW

File: src/Service.Host/client/src/components/Group.js [hunk 1]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
     };
 
+    const changeUser = e => {
+        selectUser(e.target.value);
+    };
+
+    const getEmployeeName = uri => {
+        if (members && users) {
+            const employee = users.find(x => x.href === uri);
+            if (employee) return employee.fullName;
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: could be `return employee ? employee.fullName : 'Employee not found'`. Maybe a bit clearer and you would need that second return statement below

File: src/Service.Host/client/src/components/Group.js [hunk 2]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
-                                        }
-                                        label="Active Status"
-                                        labelPlacement="start"
-                                    />
-                                </Grid>
-                            </div>
-                        </Fragment>
-                    </Grid>
+    const editGroupSection = (
+        <div className={classes.bottomPadding}>
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Is this div just being used for the padding? If so you could add that style to the Grid and remove the div

File: src/Service.Host/client/src/components/Group.js [hunk 3]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+            <Grid container spacing={24}>
+                <Grid item xs={12}>
+                    <Fragment>
+                        <div>
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: You won't need this div as you're nesting the grid items in a fragment

File: src/Service.Host/client/src/components/Group.js [hunk 4]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+        </div>
+    );
 
-    return (
+    const privilegesElements = (
+        <div className={classes.bottomPadding}>
+            <SubTitle>Privileges assigned to '{group.name}'</SubTitle>
+            {privileges && (
+                <div>
+                    <div className={classes.paddedScrollableDiv}>
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Could these divs could be removed and the styles added to the table?

File: src/Service.Host/client/src/components/Group.js [hunk 5]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                </div>
+            )}
+        </div>
+    );
+
+    const createTable = membersList => {
+        const table = [];
+        for (let i = 0; i < membersList.length; i += 4) {
+            const children = [];
+            for (let j = 0; j < 4; j += 1) {
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: I think it would be clearer to use array methods here rather than nested for loops. 4 here seems like a magic number, what's its significance? I think the reliance on the i and j variables makes this a bit hard to follow exactly what's going on.

File: src/Service.Host/client/src/components/Group.js [hunk 6]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                                <Button
+                                    type="button"
+                                   // variant="outlined"
+                                    size="small"
+                                    onClick={() =>
+                                        dispatchRemoveMember(getHref(membersList[i + j], 'self'))
+                                    }
+                                    id="RemoveMember"
+                                >
+                                    X
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Could the DeleteIcon go here?

File: src/Service.Host/client/src/components/Group.js [hunk 7]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                    >
+                        Assign to group
+                    </Button>
+                </div>
+            )}
+        </div>
+    );
+
+    const createTable = membersList => {
+        const table = [];
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Could you construct a material ui table here instead of using html elements?

File: src/Service.Host/client/src/components/Group.js [hunk 8]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                    break;
+                }
+            }
+            table.push(<tr>{children}</tr>);
+        }
+        return table;
+    };
+
+    const membersElements = (
         <div>
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Could be replaced with a fragment. Generally I think fragments are preferable for holding a bunch of react objects as divs add unneeded elements to the dom

File: src/Service.Host/client/src/components/Group.js [hunk 9]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+            <SubTitle>Members of '{group.name}'</SubTitle>
+            {members && users && (
+                <div>
+                    <ol className={classes.paddedScrollableDiv}>
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: Could be a material list, or could be replaced with the MUI table?

File: src/Service.Host/client/src/components/Group.js [hunk 10]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                        <table style={{ width: '100%' }}>{createTable(members)}</table>
+                    </ol>
+
+                    <select
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: There's a dropdown component in the shared library that might work here

File: src/Service.Host/client/src/components/Group.js [hunk 11]
```diff
@@ -46,83 +111,256 @@ const ViewGroup = ({
...
+                        value={selectedUser}
+                        defaultValue={-1}
+                        onChange={changeUser}
+                        className={classes.thinPrivilegesSelectList}
+                    >
+                        <option value={-1} key="-1">
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Group.js]: I think these -1s shouldn't be hard coded, this option won't need a custom key as its not being created in a loop

File: src/Service.Host/client/src/components/Groups.js [hunk 12]
```diff
@@ -12,18 +22,30 @@ const styles = () => ({
     }
 });
 
-const ViewGroups = ({ classes, initialise, groups, loading }) => {
+const ViewGroups = ({
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Groups.js]: These Components might be better as using the function declaration style to help debugging `function ViewGroups({ classes}) { return 'hiya' }`

File: src/Service.Host/client/src/components/Privilege.js [hunk 13]
```diff
@@ -21,19 +33,24 @@ const styles = () => ({
...
     useEffect(() => {
-        initialise();
-    }, [initialise]);
+        fetchPrivilege(id);
+      //  fetchEmployeesWithPrivilege(id),
+      fetchUsers();
+    }, [fetchPrivilege, fetchUsers, id]);
 
     const handleSaveClick = () => {
         savePrivilege(privilege.name, privilege.active, getHref(privilege, 'self'));
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Privilege.js]: There's also a getSelfHref function so you don't need to pass in 'self'

File: src/Service.Host/client/src/components/Privileges.js [hunk 14]
```diff
@@ -77,9 +78,10 @@ const ViewPrivileges = ({
...
     currentUserUri,
-    //deletePermission,
+    deletePermission,
     showPrivilegeMessage,
-    setPrivilegeMessageVisible
+    setPrivilegeMessageVisible,
+    permissionMessage
 }) => {
     useEffect(() => {
         if (selectedUser == -1) {
```
Comment (atcampbell) [inline src/Service.Host/client/src/components/Privileges.js]: Still a bit unsure about what -1 is being used for here, can you identify a user by their href?

File: src/Service/Modules/PrivilegesModule.cs [hunk 15]
```diff
@@ -41,10 +42,17 @@ private object CreatePrivilege()
             return this.Negotiate.WithModel(result);
         }
 
+        private object RemovePrivilege(int id)
+        {
+            var resource = this.Bind<PrivilegeResource>();
+            //var result = this.privilegeService.RemovePrivilege(id, resource);
+            //return this.Negotiate.WithModel(result);
+            return null;
```
Comment (atcampbell) [inline src/Service/Modules/PrivilegesModule.cs]: Something should be returned here. Maybe the user that has just been removed?

File: src/Domain/Services/PrivilegeService.cs [hunk 16]
```diff
@@ -22,11 +25,13 @@ public IEnumerable<Privilege> GetPrivileges(string who)
         {
             // TODO what if who is string.Empty ?
```
[the remark this answers is not in the document]
Comment (iaincrawf) [reply src/Domain/Services/PrivilegeService.cs]: I didn't actually write this method but will add something in!

File: src/Facade/ResourceBuilders/PermissionResourceBuilder.cs [hunk 17]
```diff
@@ -14,17 +14,19 @@ public PermissionResource Build(Permission permission)
                            {
                                GrantedByUri = individualPermission.GrantedByUri,
                                GranteeUri = individualPermission.GranteeUri,
-                               Privilege = individualPermission.Privilege.Name
-                           };
+                               Privilege = individualPermission.Privilege.Name,
+                               DateGranted = individualPermission.DateGranted.ToString("ddd, dd MMMM yyyy hh:mm tt")
```
[the remark this answers is not in the document]
Comment (iaincrawf) [reply src/Facade/ResourceBuilders/PermissionResourceBuilder.cs]: Cool will do

File: src/Service.Host/client/src/components/Group.js [hunk 18]
```diff
@@ -1,8 +1,20 @@
...
+    Grid,
+    Switch,
+    Table,
+    TableHead,
+    TableBody,
+    TableCell,
+    TableRow
+} from '@material-ui/core';
+
+// import DeleteIcon from '@material-ui/icons/Delete';
```
[the remark this answers is not in the document]
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Group.js]: Sorry, part of this stuff was mid thought and I just pushed so I could move onto the production thing with lewis - was considering using the delete icon for remove further down this

File: src/Service.Host/client/src/components/Group.js (hunk 1 above)
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Group.js]: I'd intended it to also return 'employee not found' if members or users hadn't been populated for whatever reason, so that's why I left it outside of that if statement like this, but maybe that should be a different 'error finding employee' string or something and do what you suggested inside?

File: src/Service.Host/client/src/components/Group.js (hunk 5 above)
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Group.js]: Basically wanted to put 4 names in each row of a table, but found it a pain to write this with closing tags for the <tr> etc so it's a loop that takes 4 at a time and wraps each in a <td> and the 4 in a <tr>

File: src/Service.Host/client/src/components/Privileges.js (hunk 14 above)
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Privileges.js]: in this class I used -1 to represent the "all users" option in the dropdown, and I'm using it to know if I want to get all the global privileges or just the privileges that one user actually has the permission for. I'll change it to "allUsers" or something more obvious than -1

File: src/Service/Modules/PrivilegesModule.cs (hunk 15 above)
Comment (iaincrawf) [reply src/Service/Modules/PrivilegesModule.cs]: I shouldn't have committed this as it hasn't been fully implemented yet sorry - can't remove privileges yet this was just a start towards it

File: src/Service.Host/client/src/components/Privilege.js (hunk 13 above)
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Privilege.js]: Cheers 😄

File: src/Service.Host/client/src/components/Group.js (hunk 1 above)
Comment (atcampbell) [reply src/Service.Host/client/src/components/Group.js]: Ahh yeah sorry didn't realise that

File: src/Facade/ResourceBuilders/PermissionResourceBuilder.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (iaincrawf) [reply src/Facade/ResourceBuilders/PermissionResourceBuilder.cs]: Do you print it out for users in iso format or do you convert it into something on the front end?
